packages/web/nexus/subscription.ts

```diff
@@ -0,0 +1,322 @@
+) => {
+  const customer = await stripe.customers.retrieve(customerId)
+  
+  if (!customer) throw new Error("User has stripeCustomerId but unable to find customer in Stripe")
+
+  await stripe.paymentMethods.attach(paymentMethodId, {
+    customer: customer.id,
+  })
+
+  // Update customer's default method in case they've entered a new card
+  await stripe.customers.update(customer.id, {
+    invoice_settings: {
+      default_payment_method: paymentMethodId,
+    },
+  })
+
+  return db.membershipSubscription.update({
+    where: {
+      userId,
+    },
+    data: {
+      lastFourCardNumbers,
+      cardType,
+    },
+  })
+}
+
+const setPlan = async (
+  userId: number,
+  db: PrismaClient,
+  stripeSubscriptionId: string,
+  subscriptionPeriod: MembershipSubscriptionPeriod,
+  cancelAtPeriodEnd: boolean,
+) => {
+  const stripeSubscription = await stripe.subscriptions.retrieve(stripeSubscriptionId)
+  const subscriptionUpdated = await stripe.subscriptions.update(stripeSubscription.id, {
+    payment_behavior: 'pending_if_incomplete',
+    proration_behavior: 'always_invoice',
+    items: [{
+      id: stripeSubscription.items.data[0].id,
+      price: getSubscriptionPriceId(subscriptionPeriod),
+    }],
+  })
+  if (subscriptionUpdated.pending_update) {
+    // Payment failed
+    throw new Error("Unable to update subscription, possible payment failure")
+  }
+  await stripe.subscriptions.update(stripeSubscription.id, {
+    cancel_at_period_end: cancelAtPeriodEnd,
+  })
+  // Update our records with the new subscription info
+  return db.membershipSubscription.update({
+    where: {
+      userId,
+    },
+    data: {
+      period: subscriptionPeriod,
+      expiresAt: new Date(subscriptionUpdated.current_period_end * 1000 + (24 * 60 * 60 * 1000 * 2)),
+      stripeSubscription: subscriptionUpdated as unknown as InputJsonValue,
+    },
+  })
+}
+
+const MembershipSubscriptionMutations = extendType({
+  type: 'Mutation',
+  definition(t) {
+    t.field('purchaseMembershipSubscription', {
+      type: 'MembershipSubscription',
+      args: {
+        period: arg({ type: 'MembershipSubscriptionPeriod', required: true }),
+        token: stringArg({ required: true }),
+      },
+      resolve: async (_parent, args, ctx) => {
+        const { userId } = ctx.request
+
+        if (!userId) {
+          throw new Error("You must be logged in to create a subscription")
+        }
+
+        const user = await ctx.db.user.findUnique({
+          where: {
+            id: userId,
+          },
+          include: {
+            membershipSubscription: true,
+          }
+        })
+
+        if (!user) throw new Error("User not found")
+
+        let customerId
+        if (!user.stripeCustomerId) {
+          const customer = await stripe.customers.create({
+            payment_method: args.token,
+            description: `${user.handle} (${user.id})`,
+            email: user.email,
+            metadata: {
+              id: user.id,
```

Comment:
maybe `journalyUserId` to match the metadata on the subscription?

Reply:
Great call!